### ISSUE TITLE
feat: support Android App Links

### ISSUE THREAD

**Issue by masterflitzer** [ReVanced/revanced-manager#1225]
### Type

Functionality

### Issue

going to <https://revanced.app> while already having revanced manager installed on android should open the app instead of the website

### Feature

Android App Links allow verifying a domain which is declared with an intend filter in the android manifest, android then automatically associates the domain with the app

### Motivation

this would be a quality of life improvement i guess

### Additional context

just for clarification: i tested this with my own domain and a local debug build on my phone (i just replaced revanced.app with my domain and used my debug fingerprint), when opening a link from an app like discord or from a webpage like google search the revanced manager is opened instead of website which makes linking to revanced manager directly very easy

these steps are required to implement this (more details in the android docs: <https://developer.android.com/training/app-links/verify-android-applinks>):

- AndroidManifest.xml
- branch: _compose-dev_
- autoVerify attribute for verification (needed since android 12)
- scheme: https and/or http
- host: revanced.app

```
<activity android:name=".MainActivity" android:exported="true" ...>
    ...
    <intent-filter android:autoVerify="true">
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <data android:scheme="http" />
        <data android:scheme="https" />
        <data android:host="revanced.app" />
    </intent-filter>
</activity>
```

- https://revanced.app/.well-known/assetlinks.json
- needs to be accessible without redirect (301/302/307/308)
- needs to be served as application/json
- for verification, android will check the fingerprint
- fingerprint generation (for my local debug keystore as example): `keytool -list -v -keystore ~/.android/debug.keystore`
- check verification status while debugging: `adb shell pm get-app-links app.revanced.manager`
- re-verify while debugging (check status again after): `adb shell pm verify-app-links --re-verify app.revanced.manager`
- can also be generated by this official web tool: <https://developers.google.com/digital-asset-links/tools/generator>

```json
[
   {
      "relation": [
         "delegate_permission/common.handle_all_urls"
      ],
      "target": {
         "namespace": "android_app",
         "package_name": "app.revanced.manager",
         "sha256_cert_fingerprints": [
            "00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00"
         ]
      }
   }
]
```

finally test intend manually: `adb shell am start -W -a android.intent.action.VIEW -d "https://revanced.app"`

### Acknowledgements

- [X] I have searched the existing issues and this is a new and no duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this issue properly.
- [X] The issue is related solely to the ReVanced Manager

**Comment by masterflitzer** [comment]
i don't know if i should open a PR against compose-dev or dev (i guess androidmanifest works similar in flutter) or if someone in the revanced team can do it (as they need to generate the fingerprint and deploy the json anyway)

that is if this feature gets considered/approved

**Comment by oSumAtrIX** [comment]
Going to the website should not open ReVanced Manager because it is not a replacement for the website.

**Comment by masterflitzer** [comment]
then maybe just <https://revanced.app/download>?

**Comment by oSumAtrIX** [comment]
Consider that the ReVanced Manager would have to store a constant value for the site, which is not ideal considering it should be agnostic to any.

**Comment by masterflitzer** [comment]
I understand that ReVanced should work with any repo as source for patches, but in this case it's about the download/open of the app, one should only download ReVanced manager from the official source (site or github) right?

having a constant for the same site as the default api url doesn't remove functionality from others like if you wouldn't offer changing the default api url, instead it just provides this additional functionality for the default download url

if someone wants to use a different api/repo with the manager they still will most likely download it from the current domain or github, if not and they have there own builds they could change the constant value anyway and would sign the apk with their keys and could provide their own json file with fingerprint on their domain 

I think this feature brings great comfort because at the point where you already have manager you skip the download step and jump right into the app, when someone sends you the link

also it allows to link to the app from anywhere as you normally cannot open apps from links

ultimately you guys need to decide, imo it's easy to implement, improves QoL and doesn't actually remove functionality if revanced.app is no more

**Comment by oSumAtrIX** [comment]
The point is that if someone changes the sources inside the app, the hardcoded constant wouldn't change which is a problem.

**Comment by masterflitzer** [comment]
yes you are right, it'd be nice to be able to change the url which opens the app, but on the other side this could also impose a security risk, therefore it has to be declared in the manifest and verified by domain & fingerprint, I see it this way:

the constant host should represent who compiled the app and only work for the official site where you can download it (a build from someone else would be a fork anyway and therefore they can change it easily)

while the change api url/sources is for changing the source for patches etc. while still using the official build

tldr: why should changing the source url affect the download url which opens the app? they are different concepts

**Comment by masterflitzer** [comment]
i updated the issue with the correct fingerprint (i think), seems like it can be extracted from the apk

i'd love for this feature to be implemented, maybe you can discuss it with the rest of the team?
also this would enable the revanced website itself to link to the app (with different intend filters one could link to specific activities etc.)

**Comment by oSumAtrIX** [comment]
> This could also impose a security risk

How?

---

I still see little use in this feature, considering the little to no motivation provided:

> this would be a quality of life improvement i guess

 Why is it that important to redirect? If the user ever opens the download page, if they do, regardless of having the app installed, why would they want to return to the app?

**Comment by masterflitzer** [comment]
I guess in ReVanced case it ain't a security risk, but in the general sense adjusting the value from the manifest could be a security risk because by changing some value an unverified link could be opened in an app with an malicious intent e.g. with prefilled data, this is possible because the app fingerprint is public and anybody can deploy the JSON, so the only enforcement to verify is by the intent filter in the manifest, therefore it needs to be static (I think google made it this way for a reason), but anyway we can't change that behavior anyway

regarding the actual question:
I think being able to link directly to the app improves user experience very heavily, e.g. when I make a guide for someone with step by step what to do I can say now open ReVanced manager and link to https://revanced.app/download then people either download if they don't have it already or open the manager and can do the steps

this could also be very useful for the revanced documentation by linking into the manager for convenience

**Comment by masterflitzer** [comment]
to clarify further: anybody including GitHub readme or custom made site could link to https://revanced.app/download and therefore open the manager -> very practical and convenient imho

but if you say it's unnecessary then so be it 🥲

**Comment by masterflitzer** [comment]
I agree a dialog would be nice but this would annoy people that don't already have the app (you can't detect if it's installed and only show it then)
